File: dotnet/WebService/src/WebApi/Controllers/MeasurementsController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Mime;
using AutoMapper;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using WebApi.Data;
using WebApi.Dtos;
using WebApi.Models;
using WebApi.Utils;

namespace WebApi.Controllers
{
    [Route("api/measurements")]
    [ApiController]
    public class MeasurementsController : ControllerBase
    {
        private readonly ApplicationDbContext _context;
        private readonly IMapper _mapper;

        public MeasurementsController(ApplicationDbContext context, IMapper mapper)
        {
            _context = context;
            _mapper = mapper;
        }

        [HttpGet]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status200OK)]
        public ActionResult<IEnumerable<MeasurementReadDto>> GetMeasurements([FromQuery] QueryStringParameters queryStringParameters)
        {
            int skip = (queryStringParameters.Page - 1) * queryStringParameters.Limit;
            var measurements = _context.Measurements
                .OrderBy(x => x.Id)
                .Skip(skip)
                .Take(queryStringParameters.Limit);

            var measurementReadDto = _mapper.Map<IEnumerable<MeasurementReadDto>>(measurements);

            return Ok(measurementReadDto);
        }

        [HttpGet("{id}", Name = "GetMeasurementById")]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        public ActionResult<MeasurementReadDto> GetMeasurementById([FromRoute] int id)
        {
            var measurement = _context.Measurements.FirstOrDefault(m => m.Id == id);

            if (measurement == null) return NotFound();

            var measurementReadDto = _mapper.Map<MeasurementReadDto>(measurement);

            return Ok(measurementReadDto);
        }

        [HttpPost]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status201Created)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        public ActionResult<MeasurementReadDto> CreateMeasurement([FromBody] MeasurementCreateDto measurementCreateDto)
        {
            var measurement = _mapper.Map<Measurement>(measurementCreateDto);

            _context.Measurements.Add(measurement);
            _context.SaveChanges();

            var measurementReadDto = _mapper.Map<MeasurementReadDto>(measurement);

            return CreatedAtRoute(nameof(GetMeasurementById), new { measurementReadDto.Id }, measurementReadDto);
        }

        [HttpPut("{id}")]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        public ActionResult<MeasurementReadDto> UpdateMeasurement([FromRoute] int id, [FromBody] MeasurementUpdateDto measurementUpdateDto)
        {
            var measurement = _context.Measurements.FirstOrDefault(m => m.Id == id);

            if (measurement == null) return NotFound();

            _mapper.Map(measurementUpdateDto, measurement);
            _context.SaveChanges();

            var measurementReadDto = _mapper.Map<MeasurementReadDto>(measurement);

            return Ok(measurementReadDto);
        }

        [HttpDelete("{id}")]
        [ProducesResponseType(StatusCodes.Status204NoContent)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        public ActionResult DeleteMeasurement([FromRoute] int id)
        {
            var measurement = _context.Measurements.FirstOrDefault(m => m.Id == id);

            if (measurement == null) return NotFound();

            _context.Remove(measurement);
            _context.SaveChanges();

            return NoContent();
        }

        [HttpGet("random")]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        public ActionResult<MeasurementReadDto> GetRandomMeasurement()
        {
            int randomId = Randomizer.GetNumber(1, 10001);
            var measurement = _context.Measurements.FirstOrDefault(m => m.Id == randomId);

            if (measurement == null) return NotFound();

            var measurementReadDto = _mapper.Map<MeasurementReadDto>(measurement);

            return Ok(measurementReadDto);
        }

        [HttpGet("location")]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        public ActionResult<MeasurementWithLocationReadDto> GetRandomMeasurementWithLocation()
        {
            int randomId = Randomizer.GetNumber(1, 10001);
            var measurementWithLocation = _context.Measurements
                .Include(m => m.Location)
                .FirstOrDefault(m => m.Id == randomId);

            if (measurementWithLocation == null) return NotFound();

            var measurementWithLocationReadDto = _mapper.Map<MeasurementWithLocationReadDto>(measurementWithLocation);

            return Ok(measurementWithLocationReadDto);
        }

        [HttpGet("queries")]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        public ActionResult<IEnumerable<MeasurementReadDto>> GetRandomMeasurementsByMultipleQueries([FromQuery] int count = 1)
        {
            count = (count < 1) || (count > 100) ? 1 : count;
            var measurements = new List<Measurement>();

            while (count > 0)
            {
                int randomId = Randomizer.GetNumber(1, 10001);
                var measurement = _context.Measurements.FirstOrDefault(m => m.Id == randomId);
                measurements.Add(measurement);
                count--;
            }

            measurements = measurements.OrderBy(m => m.Id).ToList();

            var measurementsReadDto = _mapper.Map<IEnumerable<MeasurementReadDto>>(measurements);

            return Ok(measurementsReadDto);
        }

        [HttpPost("random")]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        public ActionResult<MeasurementReadDto> CreateRandomMeasurement()
        {
            var timestamp = (long)DateTime.UtcNow.Subtract(DateTime.UnixEpoch).TotalSeconds;
            var measurement = new Measurement
            {
                Parameter = "pm10",
                Value = Randomizer.GetNumber(0, 101),
                Timestamp = timestamp,
                LocationId = Randomizer.GetNumber(1, 11)
            };

            _context.Measurements.Add(measurement);
            _context.SaveChanges();

            var measurementReadDto = _mapper.Map<MeasurementReadDto>(measurement);

            return CreatedAtRoute(nameof(GetMeasurementById), new { measurementReadDto.Id }, measurementReadDto);
        }

        [HttpPut("random")]
        [Produces(MediaTypeNames.Application.Json)]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        public ActionResult<MeasurementReadDto> UpdateRandomMeasurement()
        {
            int randomId = Randomizer.GetNumber(1, 10001);
            var measurement = _context.Measurements.FirstOrDefault(m => m.Id == randomId);

            if (measurement == null) return NotFound();

            var timestamp = (long)DateTime.UtcNow.Subtract(DateTime.UnixEpoch).TotalSeconds;
            measurement.Parameter = "pm10";
            measurement.Value = Randomizer.GetNumber(0, 101);
            measurement.Timestamp = timestamp;
            measurement.LocationId = Randomizer.GetNumber(1, 11);

            _context.SaveChanges();

            var measurementReadDto = _mapper.Map<MeasurementReadDto>(measurement);

            return Ok(measurementReadDto);
        }
    }
}

File: laravel/webservice/app/Http/Controllers/MeasurementsController.php
<?php

namespace App\Http\Controllers;

use App\Http\Requests\UpdateMeasurementRequest;
use App\Http\Requests\StoreMeasurementRequest;
use App\Http\Resources\MeasurementWithLocationResource;
use App\Http\Resources\MeasurementResource;
use App\Models\Measurement;
use App\Utils\Randomizer;
use Carbon\Carbon;
use Illuminate\Http\Request;

class MeasurementsController extends Controller
{
    public function index(Request $request)
    {
        $page = $request->query('page') < 1 ? 1 : $request->query('page');
        $limit = ($request->query('limit') < 1) || ($request->query('limit') > 1000) ? 10 : $request->query('limit');
        $skip = ($page - 1) * $limit;

        $measurements = Measurement::orderBy('id', 'ASC')
            ->skip($skip)
            ->take($limit)
            ->get();

        $measurementsReadDto = MeasurementResource::collection($measurements);

        return response()->json($measurementsReadDto);
    }

    public function store(StoreMeasurementRequest $request)
    {
        $validated = $request->validated();

        $measurement = new Measurement();
        $measurement->parameter = $validated['parameter'];
        $measurement->value = $validated['value'];

        if (array_key_exists('timestamp', $validated)) {
            $measurement->timestamp = $validated['timestamp'];
        } else {
            $measurement->timestamp = Carbon::now('UTC')->timestamp;
        }

        $measurement->locationId = $validated['locationId'];
        $measurement->save();

        $measurementReadDto = new MeasurementResource($measurement);

        return response()->json($measurementReadDto, 201);
    }

    public function show(Measurement $measurement)
    {
        $measurementReadDto = new MeasurementResource($measurement);

        return response()->json($measurementReadDto);
    }

    public function update(UpdateMeasurementRequest $request, Measurement $measurement)
    {
        $validated = $request->validated();

        $measurement->parameter = $validated['parameter'];
        $measurement->value = $validated['value'];

        if (array_key_exists('timestamp', $validated)) {
            $measurement->timestamp = $validated['timestamp'];
        } else {
            $measurement->timestamp = Carbon::now('UTC')->timestamp;
        }

        $measurement->locationId = $validated['locationId'];
        $measurement->save();

        $measurementReadDto = new MeasurementResource($measurement);

        return response()->json($measurementReadDto);
    }

    public function destroy(Measurement $measurement)
    {
        $measurement->delete();

        return response('', 204);
    }

    public function getRandomMeasurement()
    {
        $randomId = Randomizer::getNumber(1, 10001);
        $measurement = Measurement::findOrFail($randomId);

        $measurementReadDto = new MeasurementResource($measurement);

        return response()->json($measurementReadDto);
    }

    public function getRandomMeasurementWithLocation()
    {
        $randomId = Randomizer::getNumber(1, 10001);
        $measurementWithLocation = Measurement::with('location')->findOrFail($randomId);

        $measurementWithLocationReadDto = new MeasurementWithLocationResource($measurementWithLocation);

        return response()->json($measurementWithLocationReadDto);
    }

    public function getRandomMeasurementsByMultipleQueries(Request $request)
    {
        $count = ($request->query('count') < 1) || ($request->query('count') > 100) ? 1 : $request->query('count');

        $measurements = collect([]);

        while ($count > 0) {
            $randomId = Randomizer::getNumber(1, 10001);
            $measurement = Measurement::findOrFail($randomId);
            $measurements[] = $measurement;
            $count--;
        }

        $measurements = $measurements->sortBy('id');

        $measurementsReadDto = MeasurementResource::collection($measurements);

        return response()->json($measurementsReadDto);
    }

    public function createRandomMeasurement()
    {
        $timestamp = Carbon::now('UTC')->timestamp;
        $measurement = new Measurement();
        $measurement->parameter = 'pm10';
        $measurement->value = Randomizer::getNumber(0, 101);
        $measurement->timestamp = $timestamp;
        $measurement->locationId = Randomizer::getNumber(1, 11);
        $measurement->save();

        $measurementReadDto = new MeasurementResource($measurement);

        return response()->json($measurementReadDto, 201);
    }

    public function updateRandomMeasurement()
    {
        $randomId = Randomizer::getNumber(1, 10001);
        $measurement = Measurement::findOrFail($randomId);

        $timestamp = Carbon::now('UTC')->timestamp;
        $measurement->parameter = 'pm10';
        $measurement->value = Randomizer::getNumber(0, 101);
        $measurement->timestamp = $timestamp;
        $measurement->locationId = Randomizer::getNumber(1, 11);
        $measurement->save();

        $measurementReadDto = new MeasurementResource($measurement);

        return response()->json($measurementReadDto);
    }
}
